src/backend/optimizer/path/allpaths.c

```diff
@@ -3006,22 +3006,41 @@ recurse_push_qual(Node *setOp, Query *topquery,
 static bool
 is_query_contain_limit_groupby(Query *parse)
 {
+	ListCell   *lc;
+
+	// LIMIT, DISTINCT, GROUP BY, aggregate functions or recursive CTEs
+	// can cause motions in a plan, so return true if they are presented
 	if (parse->limitCount || parse->limitOffset ||
-		parse->groupClause || parse->distinctClause)
+		parse->groupClause || parse->distinctClause ||
+		parse->hasAggs || parse->hasRecursive)
 		return true;
 
+	// INTERSECT, EXCEPT or UNION without ALL set operations also can
```

Comment:
Why have you decided that union all can't cause motions?

Reply:
There are already test cases presented (in join.sql) with lateral and union all, they are planned ok. I couldn't find any example with UNION ALL causing motion that we need to fix. While UNION without ALL will eliminate duplicates, which is a kind of aggregation.